File: C#/task1.cs
using System;
using System.Linq;

class Program
{
    static void InputTable(double LeftPoint, double RightPoint, int N)
    {
        if (N == 1)
        {
            Console.WriteLine("N\t     a \t\t\t     b \t\t\t     b-a ");
            Console.WriteLine("----------------------------------------------------------------------------------");
        }
        Console.WriteLine($"{N} |\t{LeftPoint,10:F6}\t\t{RightPoint,10:F6}\t\t{RightPoint - LeftPoint,12:F6}");
    }

    static void InputNewtonMethod(double LeftPoint, double RightPoint, int N)
    {
        if (N == 1)
        {
            Console.WriteLine("N\t     Xn \t\t\t     Xn+1 \t\t\t     Xn+1-Xn ");
            Console.WriteLine("----------------------------------------------------------------------------------");
        }
        Console.WriteLine($"{N} |\t{LeftPoint,10:F6}\t\t{RightPoint,10:F6}\t\t{RightPoint - LeftPoint,12:F6}");
    }

    static double Function(double x)
    {
        return 2 * Math.Log(x) + 0.5 * x + 1;
    }

    static double Df(double x)
    {
        return 2 / x + 0.5;
    }

    static double ChordMethod(double a, double b, double epsilon)
    {
        int counter = 1;
        while (Math.Abs(b - a) > epsilon)
        {
            InputTable(a, b, counter);
            a = a - (b - a) * Function(a) / (Function(b) - Function(a));
            b = b - (a - b) * Function(b) / (Function(a) - Function(b));
            counter++;
        }
        Console.WriteLine("\nRoot of Method of chord = ");
        return b;
    }

    static double HalfDivisionMethod(double LeftPoint, double RightPoint, double epsilon)
    {
        int iteration = 1;
        double midPoint = 0.0;
        if (Function(LeftPoint) * Function(RightPoint) < 0)
        {
            while (Math.Abs(RightPoint - LeftPoint) > epsilon)
            {
                midPoint = (RightPoint + LeftPoint) / 2;
                InputTable(LeftPoint, RightPoint, iteration);
                if (Function(LeftPoint) * Function(midPoint) < 0)
                    RightPoint = midPoint;
                else
                    LeftPoint = midPoint;
                iteration++;
            }
        }
        else
        {
            Console.WriteLine("The interval is selected incorrectly. The function has the same signs at the ends of the segment");
        }
        Console.WriteLine("\nRoot of Half-separation method = ");
        return midPoint;
    }

    static void FindGraphicalSolution(ref float left, ref float right)
    {
        for (float x = -1; x < 5; x += 0.01f)
        {
            if (Math.Ceiling(Function(x)) == 0)
            {
                left = x - 1.0f;
                right = x + 1.0f;
            }
        }
    }

    static double NewtonMethod(double x0, double epsilon)
    {
        double x;
        for (int i = 1; Math.Abs(Function(x0)) >= epsilon && i < 10; i++)
        {
            x = x0 - Function(x0) / Df(x0);
            InputNewtonMethod(x, x0, i);
            x0 = x;
        }
        Console.WriteLine("\nRoot of Newton's method = ");
        return x0;
    }

    static void Main()
    {
        float left = 0;
        float right = 0;
        float x0 = 5;
        float eps = 0.0001f;

        FindGraphicalSolution(ref left, ref right);

        Console.WriteLine("Newton's method");
        Console.WriteLine($"{NewtonMethod(x0, eps):F6}\n");

        Console.WriteLine("Half-separation method");
        Console.WriteLine($"{HalfDivisionMethod(left, right, eps):F6}\n");

        Console.WriteLine("Method of chord");
        Console.WriteLine($"{ChordMethod(left, right, eps):F6}\n");
    }
}

File: C++/task1.cpp

#include <iostream>
#include <cmath>
#include <iomanip>
using namespace std;

void InputTable(double LeftPoint, double RightPoint, int N)
{
    if (N == 1)
    {
        cout << "N";
        cout << "\t     a ";
        cout << "\t\t\t     b ";
        cout << "\t\t\t     b-a ";
        cout << endl;
        cout << "----------------------------------------------------------------------------------";
        cout << endl;
    }
    cout << N << " |";
    cout << "\t" << setw(10) << LeftPoint;
    cout << "\t\t" << setw(10) << RightPoint;
    cout << "\t\t" << setw(12) << RightPoint - LeftPoint; //сама хорда
    cout << endl;
}
void InputNewtonMethod(double LeftPoint, double RightPoint, int N)
{
    if (N == 1)
    {
        cout << "N";
        cout << "\t     Xn ";
        cout << "\t\t\t     Xn+1 ";
        cout << "\t\t\t     Xn+1-Xn ";
        cout << endl;
        cout << "----------------------------------------------------------------------------------";
        cout << endl;
    }
    cout << N << " |";
    cout << "\t" << setw(10) << LeftPoint;
    cout << "\t\t" << setw(10) << RightPoint;
    cout << "\t\t" << setw(12) << RightPoint - LeftPoint;
    cout << endl;
}
double function(double x) {
    
    return 2 * log(x) + 0.5*x + 1; // Заменить функцией, корни которой мы ищем
}

double df(double x) 
{
    return 2/x + 0.5; // заменить на дифференциал функции, корни которой мы ищем.
}

// a, b - пределы хорды, epsilon — необходимая погрешность
double chordMethod(double a, double b, double epsilon) 
{
    int counter = 1;
    while (abs(b - a) > epsilon) // пока значение хорды больше епсилон; 
    {
        InputTable(a, b, counter);
        a = a - (b - a) * function(a) / (function(b) - function(a)); //уравнение конца а хорды (сокращение)
        b = b - (a - b) * function(b) / (function(a) - function(b));  //уравнение конца б хорды
        counter++;
    }
    // a, b — (i - 1)-й и i-й члены
    cout << "\nRoot of Method of hord = ";
    return b;
}
double HalfDivisionMethod(double LeftPoint, double RightPoint, double epsilon)
{
    int iteration = 1;
    double midPoint = 0.0;
    if (function(LeftPoint) * function(RightPoint) < 0)// проверка на разность знаков функции на концах отрезка (по условию метода)
    {
        while (abs(RightPoint - LeftPoint) > epsilon)// пока интервал больше погрешности (аналог хорд)
        {
            midPoint = (RightPoint + LeftPoint) / 2;
            InputTable(LeftPoint, RightPoint, iteration);
            if (function(LeftPoint) * function(midPoint) < 0)  RightPoint = midPoint;// если функция имеет разные знаки, то правая точка середина отрезка
            else LeftPoint = midPoint;
            //midPoint = (RightPoint + LeftPoint) / 2;
            iteration++;

            //вычисление средней точки, пока разница при сближении не станет меньше эпсилон
        }
    }
    else
    {
        cout << "The interval is selected incorrectly. The function has the same signs at the ends of the segment" << endl;
    }
    cout << "\nRoot of Half-seperation method = ";
    return midPoint;
}

void findGraficalSolution(float& left, float& right) // отделяем корни
{
    for (float x = -1; x < 5; x += 0.01) 
    {
        if (ceil(function(x)) == 0) // Свидение значений слева и справа
        {
            left = x - 1.0;
            right = x + 1.0;

        } //определение концов отрезка функции
    }
}


double NewtownMethod(double x0, double epsilon) 
{
    double x;

    for (int i = 1; abs(function(x0)) >= epsilon && i < 10; i++) //пока модуль функции больше погрешности + предел итерации
    {
        x = x0 - function(x0) / df(x0); // определение нового значения для дифиринцирования функции на касательные отрезки;  
        InputNewtonMethod(x, x0, i);
        x0 = x;
    }
    cout << "\nRoot of Newtons method = ";
    return x;
}


int main(void)
{
    //setlocale(LC_ALL, "Rus");
    float left = 0;
    float right = 0;
    float x0 = 5;
    float eps = 0.0001;

    findGraficalSolution(left, right);// отделяем корни
   // cout << left<<" "<<right<<endl;
    cout << "Newtons method" << endl;
    cout << NewtownMethod(x0, eps) << "\n" << endl;
    cout << "Half-seperation method\n" << endl;
    cout << HalfDivisionMethod(left, right, eps) << endl;
    cout << "\nMethod of hord\n" << endl;
    cout << chordMethod(left, right, eps) << endl;

    return 0;
}